events/codexClient.go

```diff
@@ -62,25 +54,51 @@ func (c *CodexClient) GetEvents(device string) []interpreter.Event {
 	return eventList
 }
 
-func (c *CodexClient) doRequest(request *http.Request) (int, []byte, error) {
-	f := circuitBreakerRequestFunc(c)
-	response, err := xhttp.RetryTransactor(c.RetryOptions, f)(request)
+func (c *CodexClient) executeRequest(request *http.Request) ([]byte, error) {
+	c.RateLimiter.Take()
+	response, err := c.CircuitBreaker.Execute(func() (interface{}, error) {
+		return c.doRequest(request)
+	})
+
 	if err != nil {
-		level.Error(c.Logger).Log(xlog.ErrorKey(), err, xlog.MessageKey(), "RetryTransactor failed")
-		return 0, []byte{}, err
+		if errors.Is(err, gobreaker.ErrOpenState) || errors.Is(err, gobreaker.ErrTooManyRequests) {
+			c.Metrics.CircuitBreakerRejectedCount.With(circuitBreakerLabel, c.CircuitBreaker.Name()).Add(1.0)
+		}
+		level.Error(c.Logger).Log(xlog.ErrorKey(), err, xlog.MessageKey(), "failed to make request")
+		return nil, err
+	}
+
+	r, ok := response.([]byte)
+	if !ok {
+		return nil, errors.New("failed to convert body to byte array")
+	}
+
+	return r, nil
+}
+
+func (c *CodexClient) doRequest(req *http.Request) (interface{}, error) {
+	if c.Metrics.RequestCount != nil {
+		c.Metrics.RequestCount.Add(1.0)
+	}
+
+	resp, err := c.Client.Do(req)
+	if resp != nil && c.Metrics.ResponseCount != nil {
+		c.Metrics.ResponseCount.With(responseCodeLabel, strconv.Itoa(resp.StatusCode)).Add(1.0)
```

Comment:
leveraging a histogram can give us interesting data oh how long codex is taking to respond with the the response code.